### PR TITLE
add framebuffer-cursormem for patching Haswell specific fCursorMemorySize

### PR DESCRIPTION
**Pull request by RehabMan**
While converting some of my projects, noticed that there was not a semantic patch for fCursorMemorySize in Haswell.

This patch adds framebuffer-cursormem for it.

### CODE REVIEW

**Comment by vit9696** [comment]
Hi,

Firstly, 1.2.3 is not released yet (and is planned somewhere in Autumn, supposedly October or November closer to 10.14.1), so this is to be included in 1.2.3, not 1.2.4.

Secondly, I think the meta side of the patch should be slightly redesigned: move the common part to a generic template function (basically leave the original function), which works on the framebuffers. Then call another framebuffer-specific function from the inside, which is empty by default and specialised for Haswell to update the cursor memory size.

I am fine with the cursor memory being an external variable in IGFX class, but comment it properly I suppose.

**Comment by RehabMan** [comment]
I'm going to need 1.2.3 for 10.14.0.
I thought it was mentioned 1.2.3 would be released soon (didn't check to verify).
Please consider releasing 1.2.3 by Mojave release, so I don't have to upload my own build.

I was just following the SNB example already in the code, but I'll see if I can find a way to share the code more effectively.

You're thinking something like:
```
	/**
	 *  Extended patching called from applyPlatformInformationListPatch
	 *
	 *  @param frame               pointer to Framebuffer data
	 *
	 */	template <typename T>
	void applyPlatformInformationPatchEx(T* frame) { }
...
// Haswell specialization of applyPlatformInformationPatchEx
template <>
void IGFX::applyPlatformInformationPatchEx(FramebufferHSW *frame) {
	// fCursorMemorySize is Haswell specific
	if (framebufferPatchFlags.bits.FPFFramebufferCursorSize) {
		frame->fCursorMemorySize = fPatchCursorMemorySize;
		DBGLOG("igfx", "fCursorMemorySize: 0x%08X", frame->fCursorMemorySize);
	}
}
...
// from applyPlatformInformationListPatch
	applyPlatformInformationPatchEx(frame);
```

??

Will fix the changelog to indicate 1.2.3.

File: WhateverGreen.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -578,7 +578,7 @@
 				MODULE_NAME = as.vit9696.WhateverGreen;
 				MODULE_START = "$(PRODUCT_NAME)_kern_start";
 				MODULE_STOP = "$(PRODUCT_NAME)_kern_stop";
-				MODULE_VERSION = 1.2.3;
+				MODULE_VERSION = 1.2.4;
```
Comment (vit9696) [inline WhateverGreen.xcodeproj/project.pbxproj]: This and below values are still pointing to the wrong version, I am afraid.

File: WhateverGreen/kern_igfx.cpp [hunk 2]
```diff
@@ -782,14 +783,21 @@ bool IGFX::applyPlatformInformationListPatch(uint32_t framebufferId, Framebuffer
...
+	}
+}
+
 template <typename T>
 bool IGFX::applyPlatformInformationListPatch(uint32_t framebufferId, T *platformInformationList) {
 	auto frame = reinterpret_cast<T *>(findFramebufferId(framebufferId, reinterpret_cast<uint8_t *>(platformInformationList), PAGE_SIZE));
 	if (!frame)
 		return false;
 
-	bool r = false;
```
Comment (vit9696) [inline WhateverGreen/kern_igfx.cpp]: I am afraid it is not the variable here being an error, but rather no use of it in the bottom where we have return true. Should probably be restored, relabled like in the other function, and the return condition should be fixed in the bottom.

File: WhateverGreen/kern_igfx.hpp [hunk 3]
```diff
@@ -110,6 +111,7 @@ class IGFX {
 	 *  Framebuffer hard-code patch
 	 */
 	FramebufferCFL framebufferPatch {};
+	uint32_t fPatchCursorMemorySize;	// Haswell only... and therefore not present in FramebufferCFL
```
Comment (vit9696) [inline WhateverGreen/kern_igfx.hpp]: Please restyle the comment like the one above. Do not want to be picky, but it is important to keep the codebase as clean as possible for own good.

File: WhateverGreen/kern_igfx.hpp [hunk 4]
```diff
@@ -412,6 +414,14 @@ class IGFX {
 	template <typename T>
 	bool applyPlatformInformationListPatch(uint32_t framebufferId, T *platformInformationList);
 
+	/**
+	 *  Extended patching called from applyPlatformInformationListPatch
+	 *
+	 *  @param frame               pointer to Framebuffer data
+	 *
+	 */	template <typename T>
```
Comment (vit9696) [inline WhateverGreen/kern_igfx.hpp]: A newline is missed here.

File: WhateverGreen/kern_igfx.cpp [hunk 5]
```diff
@@ -820,9 +830,10 @@ bool IGFX::applyPlatformInformationListPatch(uint32_t framebufferId, T *platform
 		DBGLOG("igfx", "stolenMemorySize: 0x%08X", frame->fStolenMemorySize);
 		DBGLOG("igfx", "framebufferMemorySize: 0x%08X", frame->fFramebufferMemorySize);
 		DBGLOG("igfx", "unifiedMemorySize: 0x%08X", frame->fUnifiedMemorySize);
-		r = true;
```
Comment (vit9696) [inline WhateverGreen/kern_igfx.cpp]: Ugh, I think this was deleted by accident.

Comment (RehabMan) [reply WhateverGreen/kern_igfx.cpp]: Or rather, not added back...

**Comment by RehabMan** [comment]
Should be ok now after last forced push.

FYI: My intention behind my original 'fix' for the static analyzer was to fix only the static analyzer notifications, not to fix logic bugs that may have been in the original code (original code, the function always returned true, except for the case of framebuffer id not found...)

Let me know if you want me to submit a new pull with clean/1-shot commit history.

**Comment by vit9696** [comment]
Thanks, merged =). Fortunately this place can now squash commits even from the interface, so less people have to bother.

**Comment by RehabMan** [comment]
Neat.

**Pull request merged**